src/app/icd/client/CheckInDelegate.h

```diff
@@ -39,6 +42,53 @@ class DLL_EXPORT CheckInDelegate
                                node that sent the check-in message.
      */
     virtual void OnCheckInComplete(const ICDClientInfo & clientInfo) = 0;
+
+    /**
+     * @brief Callback used to let the application know that a key refresh is
+     * needed to avoid counter rollover problems.
+     *
+     * The implementer of this function should create a new ICDRefreshKeyInfo object and a CommandSender delegate object. They
+       should also generate a new key and store it in the ICDRefreshKeyInfo object. This object will be tied to the specific key
+       refresh process and will not be used by the caller after that particular key refresh process has ended, regardless of success
+       or failure. On creating both the objects successfully, the callee should use the EstablishSession() API in ICDRefreshKeyInfo
+       to start a CASE session for re-registration.
+
+     * If the callee is unable to provide either the ICDRefreshKeyInfo or the CommandSender Delegate objects, that indicates key
+     * refresh is not possible until the callee is able to provide the required resources.
+     *
+     * @param[in] clientInfo - pointer to ICDClientInfo object representing the state associated with the
+                               node that sent the check-in message. The callee can use the clientInfo to determine the type of key
+                               to generate.
+     * @param[in] clientStorage - ICDClientStorage object stores the updated ICDClientInfo after re-registration into
+                                  persistent storage.
+     */
+    virtual void OnKeyRefreshNeeded(ICDClientInfo * clientInfo, ICDClientStorage * clientStorage) = 0;
+
+    /**
+     * @brief Callback used to let the application know that the re-registration with the new key was successful and provides the
+     * updated ICDClientInfo
+     *
+     * @param[in] clientInfo - pointer to ICDClientInfo object representing the state associated with the
+                               node that sent the check-in message. This will have the new key used for registration and the updated
+                               icd counter.
+     * @param[in] ICDRefreshKeyInfo - a pointer to the ICDRefreshKeyInfo object comprising the newly
+                                      generated key, the input ICDClientInfo, pointer to the CheckInDelegate and CommandSender
+                                      object for re-registering with the new key. The caller will NOT use icdRefreshKeyInfo passed
+                                      here anymore.
+     */
+    virtual void OnRegistrationUpdateComplete(const ICDClientInfo * clientInfo, ICDRefreshKeyInfo * icdRefreshKeyInfo) = 0;
```

Comment:
I think we don't need OnRegistrationUpdateComplete and OnRegistrationUpdateFailure.
When the check-in message is processed and re-registration for key is successful, we just need to call OnCheckInComplete, for failure case, we don't need notify the application, what do you expect from application if check-in fails?

Reply:
Updated in the new design.